mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -131,6 +131,12 @@ bool mlir::emitc::isPointerWideType(Type type) {
       type);
 }
 
+bool mlir::emitc::isSwitchOperandType(Type type) {
+  auto intType = llvm::dyn_cast<IntegerType>(type);
```

Comment:
It's fine for users to use the IfOp for boolean conditions, but that doesn't mean we should disallow using the SwitchOp. Those restrictions can make writing generic lowerings more complicated because they would need to special case for `i1`.

The `emitc` IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition. C++ doesn't require to explicitly cast to i32 (because it does it implicitly), so we shouldn't require the cast in the emitc IR either.
(It's also not so clear why e.g. i16 and i64 would be different than i8 with respect to resolveOperand).

Reply:
> The emitc IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition.

I agree with you. We will support all [types](https://en.cppreference.com/w/cpp/language/switch) for the condition. I will upload the respective patch.

Reply:
Just found that `i1` `true` converts to `-1` in `printOperation(CppEmitter &emitter, emitc::SwitchOp switchOp)`. I will fix it with a separate patch.

Reply:
Just to clarify, please add to this patch (no need to rebase, separate commit is fine) but don't address in a follow up PR. I would refer this PR as one single patch, as it needs to gets squashed at the end.

Reply:
> The emitc IR is mostly a syntactic representation of the C++ code. So I'd think that emitc.switch should allow all integral types as condition.
 
To be precise, we want to allow all integral types, opaque and index? Cases for the switch must always be integers (arg will be implicitly casted to i64). And there is no need to add the logic for parsing something like this: `case true:`, `case ENUM::...:`. Am I right?

Reply:
Would it be hard to switch to a plain ArrayAttr and parse or verify Integer, Index or Opaque attributes for the cases?

Reply:
Right now we are supporting index, opaque and integral types as arguments for the switch operator. Tests for all types we added

Reply:
Can I resolve conversation?